pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -171,46 +163,119 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
 							newRootCert := s.CA.GetCAKeyCertBundle().GetRootCertPem()
-							if !bytes.Equal(rootCert, newRootCert) {
-								rootCert = newRootCert
-								if err = ioutil.WriteFile(internalSelfSignedRootPath, rootCert, 0o600); err != nil {
-									log.Errorf("Failed to update local copy of self-signed root: %v", err)
+							if !bytes.Equal(caBundle, newRootCert) {
+								caBundle = newRootCert
+								certChain, keyPEM, err = s.CA.GenKeyCert(names, SelfSignedCACertTTL.Get(), false)
+								if err != nil {
+									log.Errorf("failed generating istiod key cert %v", err)
 								} else {
-									log.Info("Updated local copy of self-signed root")
+									s.keyCertBundleWatcher.SetAndNotify(keyPEM, certChain, caBundle)
 								}
 							}
 						}
 					}
 				}()
 				return nil
 			})
-			s.caBundlePath = internalSelfSignedRootPath
 		} else {
 			log.Infof("Use plugged-in cert at %v", signingKeyFile)
-			s.caBundlePath = path.Join(LocalCertDir.Get(), "root-cert.pem")
+			caBundle, err = ioutil.ReadFile(path.Join(LocalCertDir.Get(), "root-cert.pem"))
+			if err != nil {
+				return fmt.Errorf("failed reading %s: %v", path.Join(LocalCertDir.Get(), "root-cert.pem"), err)
+			}
 		}
-
 	} else {
 		log.Infof("User specified cert provider: %v", features.PilotCertProvider.Get())
 		return nil
 	}
-	if err != nil {
-		return err
-	}
+	s.keyCertBundleWatcher.SetAndNotify(keyPEM, certChain, caBundle)
+	return nil
+}
 
-	// Save the certificates to ./var/run/secrets/istio-dns - this is needed since most of the code we currently
-	// use to start grpc and webhooks is based on files. This is a memory-mounted dir.
-	if err := os.MkdirAll(dnsCertDir, 0o700); err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(dnsKeyFile, keyPEM, 0o600)
-	if err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(dnsCertFile, certChain, 0o600)
-	if err != nil {
-		return err
+// initCertificateWatches sets up watches for the dns certs.
+func (s *Server) initCertificateWatches(tlsOptions TLSOptions) error {
+	if hasCustomTLSCerts(tlsOptions) {
+		if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+			return fmt.Errorf("set keyCertBundle failed: %v", err)
+		}
+		// TODO: Setup watcher for root and restart server if it changes.
+		for _, file := range []string{tlsOptions.CertFile, tlsOptions.KeyFile} {
+			log.Infof("adding watcher for certificate %s", file)
+			if err := s.fileWatcher.Add(file); err != nil {
+				return fmt.Errorf("could not watch %v: %v", file, err)
+			}
+		}
+		s.addStartFunc(func(stop <-chan struct{}) error {
+			go func() {
+				var keyCertTimerC <-chan time.Time
+				for {
+					select {
+					case <-keyCertTimerC:
+						keyCertTimerC = nil
+						if err := s.keyCertBundleWatcher.SetFromFilesAndNotify(tlsOptions.KeyFile, tlsOptions.CertFile, tlsOptions.CaCertFile); err != nil {
+							log.Errorf("Setting keyCertBundle failed: %v", err)
```

Comment:
When there are failures (line 216, 227, and 229), why no error is returned?

Reply:
This is all from the legacy code, i was not thinking about fixing them in this PR.

Reply:
After a closer look, this is in a daemon monitoring thread. This is right, when it detects error,  just log an error instead of returning.